text/delegatable_cdd/README.md

```diff
@@ -0,0 +1,525 @@
+            &resp_2,
+            &challenge,
+            vec![HashSet::<usize>::new(); 2],
+            &root_issuer_vk,
+            &params1,
+            &params2,
+        )?;
+    
+    // If attributes were revealed, like in the example above, the call would be
+    let recon_com_2: AttributeTokenComm = AttributeToken::reconstruct_commitment(
+            L,
+            &com_2,
+            &resp_2,
+            &challenge,
+            vec![
+                revealed_attr_indices_1,
+                revealed_attr_indices_2,
+            ],
+            &root_issuer_vk,
+            &params1,
+            &params2,
+        )?;
+    ```
+- Commitment computation can be sped up by using precomputation of some of the
+  pairings. Precomputation can be done on the setup parameters which can be used
+  for computing commitments involving any `CredChain` or it can be done on
+  signatures as well which makes those precomputations relevant to that
+  `CredChain` only.  
+  Thus the commitment can be created using 3 methods, either
+  `AttributeToken::commitment` that does not accept any precomputation,
+  `AttributeToken::commitment_with_precomputation_on_setup_params` that only
+  accepts precomputation done on setup params (using
+  `PrecompOnSetupParams::new`) and
+  `AttributeToken::commitment_with_precomputation` that accepts precomputation
+  done on setup params (using `PrecompOnSetupParams::new`) and on `CredChain`
+  using `PrecompOnCredChain::new`
+
+    ```rust
+    // Precomputation on setup params
+    let precomputed_setup_params: PrecompOnSetupParams = PrecompOnSetupParams::new(&params1, &params2);
+
+    // Precomputation on credential chain
+    let precomp_sig_vals: PrecompOnCredChain = PrecompOnCredChain::new(&chain, &params1, &params2)?;
+
+    // Computing AttributeTokenComm using PrecompOnSetupParams
+    let at: AttributeToken = ...;
+    let com_precomp_setup: AttributeTokenComm = at
+                .commitment_with_precomputation_on_setup_params(
+                    vec![HashSet::<usize>::new(); i],
+                    &precomputed_setup_params,
+                )?
+    
+    // Computing AttributeTokenComm using PrecompOnSetupParams and PrecompOnCredChain
+    let com_precomp: AttributeTokenComm = at
+                .commitment_with_precomputation(
+                    vec![HashSet::<usize>::new(); i],
+                    &precomputed_setup_params,
+                    &precomp_sig_vals,
+                )?
+    ``` 
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Most of the implemented logic follows the paper with the corrections called out
+in the code.  
+At several places, computations like e(a, b)<sup>c</sup> have been replaced with
+e(a<sup>c</sup>, b) since they are about 30% faster. Attempts are made to
+replace computations in G2 with computations in G1 inside pairings.  
+Regarding the implementation of `verify_fast` function, it applies this
```

Comment:
@hartm @khovratovich @manudrijvers @dubovitskaya This is an implementation level optimization which was not in the paper so i wanted to check with you guys that it is fine. @hartm reviewed it and found it ok.